src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs

```diff
@@ -559,21 +559,34 @@ private void CreateTypeCastPaths(ODataPath currentPath, OpenApiConvertSettings c
                                 .OfType<IEdmStructuredType>()
                                 .ToArray();
 
-            foreach(var targetType in targetTypes)
+            foreach (var targetType in targetTypes)
             {
-                var castPath = currentPath.Clone();
                 var targetTypeSegment = new ODataTypeCastSegment(targetType);
 
+                if (currentPath.Segments.Contains(targetTypeSegment))
```

Comment:
this will always evaluate to false since the target type segment is instantiated at the previous lines (ref compare will fail) and since a comparer is not provided (value compare not done)